contrib/golang/filters/http/source/golang_filter.h

```diff
@@ -0,0 +1,260 @@
+               Logger::Loggable<Logger::Id::http>,
+               public AccessLog::Instance {
+public:
+  explicit Filter(Grpc::Context& context, FilterConfigSharedPtr config, uint64_t sid,
+                  Dso::DsoInstancePtr dynamic_lib)
+      : config_(config), dynamic_lib_(dynamic_lib), decoding_state_(*this), encoding_state_(*this),
+        context_(context), stream_id_(sid) {
+    UNREFERENCED_PARAMETER(context_);
+    UNREFERENCED_PARAMETER(stream_id_);
+  }
+
+  // Http::StreamFilterBase
+  void onDestroy() ABSL_LOCKS_EXCLUDED(mutex_) override;
+  Http::LocalErrorStatus onLocalReply(const LocalReplyData&) override;
+
+  // Http::StreamDecoderFilter
+  Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap& headers,
+                                          bool end_stream) override;
+  Http::FilterDataStatus decodeData(Buffer::Instance&, bool) override;
+  Http::FilterTrailersStatus decodeTrailers(Http::RequestTrailerMap&) override;
+  void setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callbacks) override {
+    decoding_state_.setDecoderFilterCallbacks(callbacks);
+  }
+
+  // Http::StreamEncoderFilter
+  Http::Filter1xxHeadersStatus encode1xxHeaders(Http::ResponseHeaderMap&) override {
+    return Http::Filter1xxHeadersStatus::Continue;
+  }
+  Http::FilterHeadersStatus encodeHeaders(Http::ResponseHeaderMap& headers,
+                                          bool end_stream) override;
+  Http::FilterDataStatus encodeData(Buffer::Instance& data, bool end_stream) override;
+  Http::FilterTrailersStatus encodeTrailers(Http::ResponseTrailerMap& trailers) override;
+  Http::FilterMetadataStatus encodeMetadata(Http::MetadataMap&) override {
+    return Http::FilterMetadataStatus::Continue;
+  }
+
+  void setEncoderFilterCallbacks(Http::StreamEncoderFilterCallbacks& callbacks) override {
+    encoding_state_.setEncoderFilterCallbacks(callbacks);
+  }
+
+  // AccessLog::Instance
+  void log(const Http::RequestHeaderMap* request_headers,
+           const Http::ResponseHeaderMap* response_headers,
+           const Http::ResponseTrailerMap* response_trailers,
+           const StreamInfo::StreamInfo& stream_info) override;
+
+  void onStreamComplete() override;
+
+  // create metadata for golang.extension namespace
+  void addGolangMetadata(const std::string& k, const uint64_t v);
+
+  static std::atomic<uint64_t>& getGlobalStreamId() {
+    MUTABLE_CONSTRUCT_ON_FIRST_USE(std::atomic<uint64_t>, 0);
+  }
+
+  void continueStatus(GolangStatus status);
+
+  void sendLocalReply(Http::Code response_code, absl::string_view body_text,
+                      std::function<void(Http::ResponseHeaderMap& headers)> modify_headers,
+                      Grpc::Status::GrpcStatus grpc_status, absl::string_view details);
+
+  absl::optional<absl::string_view> getHeader(absl::string_view key);
+  void copyHeaders(GoString* go_strs, char* go_buf) ABSL_LOCKS_EXCLUDED(mutex_);
```

Comment:
You do not want to be using `ABSL_LOCKS_EXCLUDED`. Remove all of this. You want to be using `ABSL_EXCLUSIVE_LOCKS_REQUIRED` on the variables that need to be locked, and then on methods where the lock is already acquired if applicable. Same elsewhere.

Reply:
@mattklein123 You mean to delete the `lock` in the `copyHeaders` and add `ABSL_EXCLUSIVE_LOCKS_REQUIRED` annotation to it. Finally use lock before calling this function?
https://github.com/envoyproxy/envoy/blob/c063d61d52c81013e4620b4ddfedf7f196289de8/contrib/golang/filters/http/source/golang_filter.cc#L658-L659

Reply:
Done. add ABSL_EXCLUSIVE_LOCKS_REQUIRED for some methods.